### PR TITLE
added pixel-example.cc

### PR DESCRIPTION
**Pull request by punyanagari**
Display a single pixel with any colour and brightness.
Move around the pixel with W,A,S,D keys.
Helpful to test/debug newer panels with unknown settings.

### CODE REVIEW

File: examples-api-use/pixel-example.cc [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        break;
+     case 'd':
+	if(x_pos < canvas->width() - 1)
+          x_pos++;
+        break;
+    }
+  }
+  // Finished. Shut down the RGB matrix.
+  canvas->Clear();
+  delete canvas;
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:146]: Could be done in handler

File: examples-api-use/pixel-example.cc [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  int x_pos=0;
+  int y_pos=0;
+
+  fprintf(stderr, "Move around with W,A,S,D keys.\n"
+                  "The pixel position cannot be less than 0 and greater than the display height or width.\n");
+
+  while (!interrupt_received) {
+    canvas->Clear();
+    canvas->SetPixel(x_pos,y_pos, color.r, color.g, color.b);
+    fprintf(stderr, "X,Y Position : %d,%d\n",x_pos,y_pos);   //display the pixel position
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:122]: Print old and new position?

File: examples-api-use/pixel-example.cc [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  int x_pos=0;
+  int y_pos=0;
+
+  fprintf(stderr, "Move around with W,A,S,D keys.\n"
+                  "The pixel position cannot be less than 0 and greater than the display height or width.\n");
+
+  while (!interrupt_received) {
+    canvas->Clear();
+    canvas->SetPixel(x_pos,y_pos, color.r, color.g, color.b);
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:121]: Need to be done every pass?

File: examples-api-use/pixel-example.cc [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    fprintf(stderr, "X,Y Position : %d,%d\n",x_pos,y_pos);   //display the pixel position
+    char c;
+    c=getch();
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:124]: Need to set attributes every time or just read?

File: examples-api-use/pixel-example.cc (hunk 2 above)
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:122]: stderr?

File: examples-api-use/pixel-example.cc [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  if (all_extreme_colors)
+    canvas->SetPWMBits(1);
+
+  signal(SIGTERM, InterruptHandler);
+  signal(SIGINT, InterruptHandler);
+
+  int x_pos=0;
+  int y_pos=0;
+
+  fprintf(stderr, "Move around with W,A,S,D keys.\n"
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:116]: stderr?

File: examples-api-use/pixel-example.cc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        fprintf(stderr, "Invalid color spec: %s\n", optarg);
+        return usage(argv[0]);
+      }
+      break;
+    default:
+      return usage(argv[0]);
+    }
+  }
+  if (brightness < 1 || brightness > 100) {
+    fprintf(stderr, "Brightness is outside usable range.\n");
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:94]: stderr?

File: examples-api-use/pixel-example.cc [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  Color color(255, 255, 0);
+  int brightness = 100;
+
+  int opt;
+  while ((opt = getopt(argc, argv, "C:b:")) != -1) {
+    switch (opt) {
+    case 'b': brightness = atoi(optarg); break;
+    case 'C':
+      if (!parseColor(&color, optarg)) {
+        fprintf(stderr, "Invalid color spec: %s\n", optarg);
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:85]: stderr?

File: examples-api-use/pixel-example.cc [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  interrupt_received = true;
+}
+
+static int usage(const char *progname) {
+  fprintf(stderr, "usage: %s [options]\n", progname);
+  fprintf(stderr, "Display single pixel with any colour and brightness.\n"
+		  "Move around with W,A,S,D keys.\n"
+   		  "The pixel position cannot be less than 0 or greater than the display height and width.\n");
+  fprintf(stderr, "Options:\n\n");
+  fprintf(stderr,
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:50]: stderr?

File: examples-api-use/pixel-example.cc [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        old.c_cc[VMIN] = 1;
+        old.c_cc[VTIME] = 0;
+        if (tcsetattr(0, TCSANOW, &old) < 0)
+                perror("tcsetattr ICANON");
+        if (read(0, &buf, 1) < 0)
+                perror ("read()");
+        old.c_lflag |= ICANON;
+        old.c_lflag |= ECHO;
+        if (tcsetattr(0, TCSADRAIN, &old) < 0)
+                perror ("tcsetattr ~ICANON");
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:33]: Does not restore state properly. Is restore needed?

File: examples-api-use/pixel-example.cc [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+char getch() {
+        char buf = 0;
+        struct termios old;
+        if (tcgetattr(0, &old) < 0)
+                perror("tcsetattr()");
+        old.c_lflag &= ~ICANON;
+        old.c_lflag &= ~ECHO;
+        old.c_cc[VMIN] = 1;
+        old.c_cc[VTIME] = 0;
+        if (tcsetattr(0, TCSANOW, &old) < 0)
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:26]: Needed to be done every time?

File: examples-api-use/pixel-example.cc [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        if (tcsetattr(0, TCSADRAIN, &old) < 0)
+                perror ("tcsetattr ~ICANON");
+        return (buf);
+}
+
+using namespace rgb_matrix;
+
+volatile bool interrupt_received = false;
+static void InterruptHandler(int signo) {
+  interrupt_received = true;
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:41]: Need flag? Could make canvas pointer global and clean up here. (Null checks)

File: examples-api-use/pixel-example.cc [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  signal(SIGTERM, InterruptHandler);
+  signal(SIGINT, InterruptHandler);
+
+  int x_pos=0;
+  int y_pos=0;
+
+  fprintf(stderr, "Move around with W,A,S,D keys.\n"
+                  "The pixel position cannot be less than 0 and greater than the display height or width.\n");
+
+  while (!interrupt_received) {
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:119]: Not sure the flag is needed

File: examples-api-use/pixel-example.cc [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+#include <signal.h>
+#include <termios.h>
+
+char getch() {
+        char buf = 0;
+        struct termios old;
+        if (tcgetattr(0, &old) < 0)
+                perror("tcsetattr()");
+        old.c_lflag &= ~ICANON;
+        old.c_lflag &= ~ECHO;
```
Comment (daveythacher) [inline examples-api-use/pixel-example.cc:23]: Is this really this important?

Comment (daveythacher) [reply examples-api-use/pixel-example.cc:23]: Plan to script this?

**Comment by daveythacher** [comment]
Cool program

**Comment by hzeller** [comment]
Thanks! Merged in https://github.com/hzeller/rpi-rgb-led-matrix/commit/3ddf4e4f300c37525413d4979c7d00a05398e178

(I only had to fix some indentations to use the 2 space indent used in this project)

**Comment by daveythacher** [comment]
Everything is printed to stderr. Usage, error messages, general messages, etc. Kind of what I expected.

File: examples-api-use/pixel-example.cc (hunk 11 above)
Comment (hzeller) [reply examples-api-use/pixel-example.cc:41]: No, you can't. You must not do anything complex in a signal handler.

Comment (daveythacher) [reply examples-api-use/pixel-example.cc:41]: Yeah these are not interrupt handles. Granted this suck pretty bad. In all fairness this is probably not needed at all.

**Pull request closed without merging**